server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -664,7 +664,12 @@ public Set<String> resolveExpressions(ClusterState state, String... expressions)
             getSystemIndexAccessPredicate(),
             getNetNewSystemIndexPredicate()
         );
-        return Set.copyOf(resolveExpressions(Arrays.asList(expressions), context));
+        Collection<String> resolved = resolveExpressions(Arrays.asList(expressions), context);
+        if (resolved instanceof Set<String>) {
```

Comment:
I think this change slightly changes the semantics. It appears that the underlying resolve methods can return a mutable Set, while previously with Set.copyOf we would always return unmodifiable set. Since this is a public method, I think we should retain the immutable semantics.

Slight improvement to the code, with the instanceof expressions now we can avoid the cast on the return:

```
if (resolved instanceof Set<String> resolvedSet) {
   return Set.copyOf(resolvedSet);
}

Reply:
This has the potential to be a long list; there are benchmarks where this is 50k. I've preferred to avoid the copying.
If we want to maintain the unmodifiable semantics, should we instead use `Collections.unmodifiableSet`, as that doesn't copy the Set?

Reply:
Yes, thanks using an unmodifiableSet would work great.

Reply:
I have pushed https://github.com/elastic/elasticsearch/pull/89311/commits/da5565a2aa07c73ba9aeec23f7a15110f1f5da75  . I think ideally the return result of `resolveExpressions` should be unmodifiable, ie without making it so at this level, but maybe I'll tackle that later.